File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/cli/verbs/optimize.mjs
import { statSync } from "node:fs";
import { resolve, join } from "node:path";
import { optimize } from "../../optimize/index.mjs";
import { applyVariant, commit, rollback } from "../../optimize/apply.mjs";
import { validate } from "./validate.mjs";

function parseArgs(args) {
  const opts = {
    path: null,
    format: "human",
    baselineOnly: false,
    apply: null,
    queryCount: 8,
  };
  for (let i = 0; i < args.length; i++) {
    const a = args[i];
    if (a === "--baseline-only") opts.baselineOnly = true;
    else if (a === "--apply") {
      const v = args[++i];
      const n = parseInt(v, 10);
      if (!Number.isNaN(n)) opts.apply = n;
    } else if (a === "--format") opts.format = args[++i];
    else if (a === "--query-count") {
      const v = args[++i];
      const n = parseInt(v, 10);
      if (!Number.isNaN(n) && n > 0) opts.queryCount = n;
    } else if (!a.startsWith("--") && opts.path == null) opts.path = a;
  }
  return opts;
}

function sortKeys(v) {
  if (Array.isArray(v)) return v.map(sortKeys);
  if (v && typeof v === "object") {
    const o = {};
    for (const k of Object.keys(v).sort()) o[k] = sortKeys(v[k]);
    return o;
  }
  return v;
}

function formatHuman(report) {
  const lines = [];
  lines.push(`${report.path}`);
  lines.push(`  provider: ${report.provider}`);
  const b = report.baseline;
  lines.push(
    `  baseline: score=${b.score}/100 hit=${b.hit_rate.positive_hits}/${b.hit_rate.positive_total} false=${b.hit_rate.false_positives}/${b.hit_rate.false_total}`,
  );
  if (report.variants.length > 0) {
    lines.push("  variants:");
    report.variants.forEach((v, i) => {
      lines.push(
        `    [${i}] score=${v.score}/100 hit=${v.hit_rate.positive_hits}/${v.hit_rate.positive_total} false=${v.hit_rate.false_positives}/${v.hit_rate.false_total}`,
      );
      lines.push(`        rationale: ${v.rationale}`);
      lines.push(`        text: ${v.text}`);
    });
  }
  return lines.join("\n") + "\n";
}

export async function optimizeVerb(args) {
  const opts = parseArgs(args);
  if (!opts.path) {
    return {
      exitCode: 64,
      output:
        "usage: skills-ref optimize <skill-dir> [--baseline-only] [--apply N] [--format human|json] [--query-count N]\n",
    };
  }
  const target = resolve(opts.path);
  let st;
  try {
    st = statSync(target);
  } catch {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }
  if (!st.isDirectory()) {
    return { exitCode: 2, output: `not a directory: ${opts.path}\n` };
  }
  try {
    statSync(join(target, "SKILL.md"));
  } catch {
    return { exitCode: 2, output: `no SKILL.md found under: ${opts.path}\n` };
  }

  let report;
  try {
    report = await optimize(target, {
      baselineOnly: opts.baselineOnly,
      queryCount: opts.queryCount,
    });
  } catch (e) {
    return { exitCode: 1, output: `optimize: ${e.message}\n` };
  }

  if (opts.apply != null) {
    const idx = opts.apply;
    if (idx < 0 || idx >= report.variants.length) {
      return {
        exitCode: 64,
        output: `--apply ${idx}: variant index out of range (have ${report.variants.length})\n`,
      };
    }
    const variant = report.variants[idx];
    let backup;
    try {
      ({ backup } = applyVariant(target, variant));
    } catch (e) {
      return { exitCode: 1, output: `optimize --apply: ${e.message}\n` };
    }
    const v = await validate([
      target,
      "--extra-tiers",
      "company,enterprise,application",
      "--format",
      "json",
    ]);
    if (v.exitCode !== 0) {
      rollback(target, backup);
      return {
        exitCode: 1,
        output: `optimize --apply ${idx}: post-write validation failed; original restored.\n`,
      };
    }
    commit(backup);
    return {
      exitCode: 0,
      output: `optimize --apply ${idx}: wrote variant; validation OK.\n`,
    };
  }

  let output;
  if (opts.format === "json") {
    output = JSON.stringify(sortKeys(report)) + "\n";
  } else {
    output = formatHuman(report);
  }
  return { exitCode: 0, output };
}

File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/verbs/optimize.py
from __future__ import annotations

import json
from pathlib import Path

from ..optimize.apply import apply_variant, commit, rollback
from ..optimize.orchestrator import optimize as _optimize
from .validate import validate as _validate


def _sort(v):
    if isinstance(v, list):
        return [_sort(x) for x in v]
    if isinstance(v, dict):
        return {k: _sort(v[k]) for k in sorted(v.keys())}
    return v


def _format_human(report: dict) -> str:
    lines = [report["path"], f"  provider: {report['provider']}"]
    b = report["baseline"]
    hr = b["hit_rate"]
    lines.append(
        f"  baseline: score={b['score']}/100 hit={hr['positive_hits']}/{hr['positive_total']} false={hr['false_positives']}/{hr['false_total']}"
    )
    if report["variants"]:
        lines.append("  variants:")
        for i, v in enumerate(report["variants"]):
            vh = v["hit_rate"]
            lines.append(
                f"    [{i}] score={v['score']}/100 hit={vh['positive_hits']}/{vh['positive_total']} false={vh['false_positives']}/{vh['false_total']}"
            )
            lines.append(f"        rationale: {v['rationale']}")
            lines.append(f"        text: {v['text']}")
    return "\n".join(lines) + "\n"


class _Ns:
    def __init__(self, path, fmt, extra_tiers, no_color=False, strict=False):
        self.path = path
        self.format = fmt
        self.extra_tiers = extra_tiers
        self.no_color = no_color
        self.strict = strict


def optimize(args) -> tuple[int, str]:
    target = Path(args.path)
    if not target.exists():
        return 2, f"no SKILL.md found under: {args.path}\n"
    if not target.is_dir():
        return 2, f"not a directory: {args.path}\n"
    if not (target / "SKILL.md").is_file():
        return 2, f"no SKILL.md found under: {args.path}\n"

    try:
        report = _optimize(
            str(target.resolve()),
            {
                "baseline_only": bool(getattr(args, "baseline_only", False)),
                "query_count": int(getattr(args, "query_count", 8) or 8),
            },
        )
    except Exception as e:
        return 1, f"optimize: {e}\n"

    apply_idx = getattr(args, "apply", None)
    if apply_idx is not None:
        if apply_idx < 0 or apply_idx >= len(report["variants"]):
            return (
                64,
                f"--apply {apply_idx}: variant index out of range (have {len(report['variants'])})\n",
            )
        variant = report["variants"][apply_idx]
        try:
            res = apply_variant(str(target.resolve()), variant)
        except Exception as e:
            return 1, f"optimize --apply: {e}\n"
        backup = res["backup"]
        v_args = _Ns(
            path=str(target.resolve()),
            fmt="json",
            extra_tiers="company,enterprise,application",
        )
        exit_code, _ = _validate(v_args)
        if exit_code != 0:
            rollback(str(target.resolve()), backup)
            return (
                1,
                f"optimize --apply {apply_idx}: post-write validation failed; original restored.\n",
            )
        commit(backup)
        return 0, f"optimize --apply {apply_idx}: wrote variant; validation OK.\n"

    if args.format == "json":
        return 0, json.dumps(_sort(report), separators=(",", ":")) + "\n"
    return 0, _format_human(report)
